Annotation model

Allcharles opened the pull request:
# Implemented Annotation Pages

Created the my annotations and their annotations pages

## Changes

- Implemented My Annotation Page
- Implemented Their Annotation Page
- Modified how the Association Decorators function slightly to reduce the number of excess queries, and improve type checking
- Created pipe for displaying loading animation
- Created pipe for testing if a model is resolved
- Created pipe for converting `Duration`s and `DateTime`s to strings

## Problems

- [x] Need to create dummy links for Play and Visualize buttons
- [x] Need to update unit tests
- [x] Modifications to UnresolvedModel.many should be reverted. Checking for a list of UnresolvedModels can be done using a comparison of pointers

## Issues

Fixes: #138
Fixes: #563 

## Visual Changes

### Page Loading

![temp](https://user-images.githubusercontent.com/3955116/97836695-9162ca00-1d28-11eb-9159-ed04fd2ced16.gif)

### My Annotations Page

![image](https://user-images.githubusercontent.com/3955116/97667292-5b74da00-1acb-11eb-83fe-125a4f58df65.png)

### Their Annotations Page

![image](https://user-images.githubusercontent.com/3955116/97667375-747d8b00-1acb-11eb-9957-993e5a6e65ac.png)


## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing

Allcharles committed:
Cleanup route

Forgot to change back the analysis job view url

Allcharles committed:
Attempt to filter recordings by audio event creator id

Allcharles committed:
Initial look at annotations table

Allcharles committed:
Commit code across computers

Allcharles committed:
Cleanup AssociationDecorator types

Cosmetic changes to ServiceToken to attempt to clean them up

Allcharles committed:
Working my/their-annotations page

Modified the many UnresolvedModels array to contain one UnresolvedModel. This maintains the output as an array, whilst also giving the ability to determine if the value is still loading.
Would like to make modifications to the HasMany decorator so that it short circuits requests which contain no ids

Allcharles committed:
Reduce HasMany excess queries

HasMany was making requests to the API even when it had an empty array of ids. This has now been prevented by making it terminate early and return the failureValue

Allcharles committed:
Updated table decriptions

Updated the description above paged table template using components so that they will display the number of models in the current page, and the total number of models.
Discoved admin tag groups list page is broken. Api endpoint is wrong

Allcharles commented on src/app/helpers/tableTemplate/pagedTableTemplate.ts:
```diff
@@ -152,7 +154,7 @@ export abstract class PagedTableTemplate<T, M extends AbstractModel>
   }
 
-  protected apiAction(filters: Filters<M>, args: AbstractModel[]) {
+  protected apiAction(filters: Filters<M>, args?: AbstractModel[]) {
```
I suspect this should instead be `protected apiAction(filters: Filters<M>, args: AbstractModel[] = []) {`

atruskie replied:
I don't think it matters. According to the spec it should resolve the same when its spread.

atruskie commented on src/app/components/profile/pages/annotations/annotations.component.html:
```diff
@@ -21,11 +22,35 @@ <h1>
     <ngx-datatable-column name="Site" [sortable]="false">
       <ng-template let-value="value" ngx-datatable-cell-template>
-        <a [routerLink]="[value.viewUrl]">{{ value.name }}</a>
+        {{ value.audioRecording?.site?.name || "(loading)" }}
```
make a loading standard pipe?

atruskie commented on src/app/components/profile/pages/annotations/annotations.component.html:
```diff
@@ -21,11 +22,35 @@ <h1>
+      <ng-template let-value="value" ngx-datatable-cell-template>
+        <ng-container *ngFor="let tag of value.tags">
+          <span class="badge badge-primary mr-1">
+            {{ tag.text || "(loading)" }}
```
pipe?

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -101,7 +101,7 @@ export abstract class AbstractModel {
 export class UnresolvedModel extends AbstractModel {
   private static readonly model = Object.freeze(new UnresolvedModel());
-  private static readonly models = [];
+  private static readonly models = [UnresolvedModel.model];
```
should freexe this array to so it can't be mutated

atruskie commented on src/app/models/AssociationDecorators.ts:
```diff
@@ -1,205 +1,224 @@
+ *
+ * @param serviceToken Injection token for API service used to retrieve the child models
+ * @param identifierKeys Parent model key used to retrieve the list of ids for the child models
+ * @param childIdentifier Primary key of the child model
```
```suggestion
 * @param childIdentifier field used to filter child models by, typically ìd`
```

atruskie commented on src/app/models/AudioEvent.ts:
```diff
@@ -50,14 +65,23 @@ export class AudioEvent extends AbstractModel implements IAudioEvent {
     super(audioEvent, injector);
+    this.taggings = (audioEvent.taggings as ITagging[]).map(
+      (tagging) => new Tagging(tagging, injector)
+    );
```
As an FYI this pattern is very rare. Our API advice acturally states whole nested objects should never be serialised

Allcharles committed:
Created isUnresolved pipe

Allcharles committed:
Created helper pipes

Allcharles committed:
Update src/app/models/AssociationDecorators.ts

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>

Allcharles committed:
Merge branch 'annotation-model' of https://github.com/QutEcoacoustics/workbench-client into annotation-model

Allcharles committed:
Making tests compilable

Allcharles committed:
Implemented pipe tests

Allcharles committed:
Updated AssociationDecorator tests

Breaking change to how association decorators work. They no longer accept multiple responses because it has no use case, and complicates an already complicated function. This logic can be easily re-implemented. Tests for this scenario have been removed.

Allcharles committed:
Cleanup remaining tests

Allcharles committed:
Implemented dummy links

Allcharles committed:
Merge branch 'master' into annotation-model

atruskie commented on src/app/models/AssociationDecorators.ts:
```diff
@@ -45,7 +45,7 @@ export function Deleter<Parent extends AbstractModel & { deleterId?: Id }>() {
  * @param serviceToken Injection token for API service used to retrieve the child models
  * @param identifierKeys Parent model key used to retrieve the list of ids for the child models
- * @param childIdentifier Primary key of the child model
+ * @param childIdentifier field used to filter child models by, typically ìd`
```
Looks like by suggestion was bad: I has the international keyboard on!
```suggestion
 * @param childIdentifier field used to filter child models by, typically `id`
```

atruskie commented on src/app/components/profile/pages/annotations/annotations.component.html:
```diff
@@ -22,7 +22,9 @@ <h1>
       <ng-template let-value="value" ngx-datatable-cell-template>
-        {{ value.audioRecording?.site?.name || "(loading)" }}
+        <span
+          [innerHTML]="value.audioRecording?.site?.name | spinner: 'sm'"
```
when you get to the point of modifying html, it is probably time for a component?

atruskie reviewed (approved):
I think maybe the spinner pipe is best collapsed into the spinner component we have.

Other than that, great!

Allcharles committed:
Re-implement loading spinner

Allcharles committed:
Remove spinner pipe
